cupy/_manipulation/kind.py

```diff
@@ -2,6 +2,30 @@
 from cupy import _core
 
 
+def asarray_chkfinite(a, dtype=None, order=None):
+    """Converts the given input to an array,
+    and raises an error if the input contains NaNs or Infs.
+
+    Args:
+        a: array like.
+        dtype: data type, optional
+        order: {'C', 'F', 'A', 'K'}, optional
+
+    Returns:
+        cupy.ndarray: An array on the current device.
+
+    .. seealso:: :func:`numpy.asarray_chkfinite`
+
+    """
+
+    a = cupy.asarray(a, dtype=dtype, order=order)
```

Comment:
To avoid synchronization as much as possible, I think it's better to do inf check on CPU.

```py
if isinstance(a, cupy.ndarray):
    if not cupy.isfinite(a).all():
        raise ValueError(...)
else:
    a = cupy.asarray(np.asarray_chkfinite(a, dtype=dtype, order=order), dtype=dtype, order=order))
return a
```

Reply:
@kmaehashi Does it matter to avoid synchronization here if an exception is going to be raised anyway? Wouldn't it be better to raise sooner (by doing the check on GPU)? Put it differently, if the check passes (no exception) this takes longer than the original implementation.

Reply:
Sorry my proposal was inaccurate, this is what I intended:
```py
if isinstance(a, cupy.ndarray):
    if not cupy.isfinite(a).all():
        raise ValueError(...)
else:
    a = np.asarray_chkfinite(a, dtype=dtype, order=order)
return cupy.asarray(a, dtype=dtype, order=order)
```

When an array is `cupy.ndarray`, there's no difference between the original implementation and my proposal. When an array is CPU array-like, the original implementation copies the array to GPU, checks for inf on GPU, and copies the result back to CPU. My proposal changes this to check for inf on CPU and (if no exceptions are raised) copy the array it to GPU.

Reply:
From the above discussion:
#### Original Implementation for CPU array
- Copies the arrays from CPU to GPU
- Perform Operations on GPU
- Copies back to CPU
#### Your Implementation for CPU array
- Performs operations on CPU
- Copies to GPU (if there are no infs and nans)
#### As per @leofang's comment
- If checks are passed, the implementation takes a longer time.
- GPU error's will be faster

@leofang, @kmaehashi, I have a small doubt. Like in the original implementation, we are making two implicit switches (CPU->GPU->CPU), while by your proposal, we are making a single switch (CPU->GPU). I am not too sure, but don't you think these *implicit* switches will affect the model's performance? Which factor to consider prior?

Thanks for the discussions! (:

Reply:
> When an array is `cupy.ndarray`, there's no difference between the original implementation and my proposal.

Yes.

> When an array is CPU array-like, the original implementation copies the array to GPU, checks for inf on GPU, and copies the result back to CPU. My proposal changes this to check for inf on CPU and (if no exceptions are raised) copy the array it to GPU.

Depends on our expectation for the likelihood of raising an error (due to either inf or nan) for this routine, the decision could differ. @kmaehashi has implicitly assumed that it's more likely to raise, whereas I expected in most cases the check will pass. In the former case, it's OK to do the check on host (which is slower) since the subsequent control flow will be returned to host, while in the latter case, it's cheaper to run the check on device (faster) and proceed to the next GPU calculation right away if there's no error.

Reply:
Hmm I can be missing something.

> it's cheaper to run the check on device (faster) and proceed to the next GPU calculation right away

The evaluation of `if not cupy.isfinite(a).all()` always requires a copy of 0-dim bool array from device to host, so synchronization happens in any case (whether the inf check pass or not).

> I expected in most cases the check will pass

I also expect this :)

Reply:
Yes, but If the check runs much faster on GPU than on CPU, then it's probably ok to pay the sync cost 🙂

Reply:
cc: @kmaehashi @leofang

As per the above discussion, I updated the function considering speed as a priority.
Please let me know if anything else is required. I'll be happy to work on it.